## 3.6 Homework

Previous homeworks:

**Week 1:**
* https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/01-intro/homework.md

**Week 2:**
* https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/02-experiment-tracking/homework.md

The goal of this homework is to familiarize users with workflow orchestration. We start from the solution of homework 1. The notebook can be found below:

https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/01-intro/homework.ipynb

This has already been converted to a script called `homework.py` in the `03-orchestration` folder of this repo.

## Motivation

We already have a model training script. Maybe a data scientist in your team handed it to you and your job is schedule the running of training script using a workflow orchestration - Prefect in this case. Below are the requirements. Do not implement them yet, we will do so in this exercise. Just understand the goal.

1. The training flow will be run every month.
2. The flow will take in a parameter called `date` which will be a datetime.
    a. `date` should default to None
    b. If `date` is None, set `date` as the current day. Use the data from 2 months back as the training data and the data from the previous month as validation data.
    c. If `date` is passed, get 2 months before the `date` as the training data, and the previous month as validation data.
    d. As a concrete example, if the date passed is "2021-03-15", the training data should be "fhv_tripdata_2021-01.parquet" and the validation file will be "fhv_trip_data_2021-02.parquet"
3. Save the model as "model-{date}.pkl" where date is in `YYYY-MM-DD`. Note that `date` here is the value of the flow `parameter`. In practice, this setup makes it very easy to get the latest model to run predictions because you just need to get the most recent one.
4. In this example we use a DictVectorizer. That is needed to run future data through our model. Save that as "dv-{date}.pkl". Similar to above, if the date is `2021-03-15`, the files output should be `model-2021-03-15.bin` and `dv-2021-03-15.b`.

This convention is not strict in industry, and in practice, you will come up with your own system to manage these training pipeline runs. For example, if we wanted to train on the whole history instead of just one month, we'd need to allow for added parameterization and logic in our flow. If the data came in weekly instead of monthly, we might need a different naming convention. But these requirements are already a simple approximation of something you could use in production.

On the deployment side, it's very easy to just pull in the latest data and predict it using the latest model and vectorizer files. Tools the MLFlow in the last chapter can simplify that process as well. This homework will focus more on the batch training.

In order, this homework assignment will be about:

1. Converting the script to a Flow
2. Changing the parameters to take in a `date`. Making this parameter dynamic.
3. Scheduling a batch training job that outputs the latest model somewhere

## Setup

You can use either local Prefect Orion or a VM hosted Prefect Orion instance for this. It shouldn't matter. Just note that if you use a VM hosted one, you will need to configure your local API to hit the VM.

Video 3.4 of the course will give more detailed instructions if you been run it on a VM.

## Get the data

Before running anything at all, get all trip data for January-March 2021

In [1]:
# use brace expansion to download all January - March files
!wget -P data 'https://nyc-tlc.s3.amazonaws.com/trip+data/fhv_tripdata_2021-0'{1..3}'.parquet'

--2022-06-04 23:19:14--  https://nyc-tlc.s3.amazonaws.com/trip+data/fhv_tripdata_2021-01.parquet
Resolving nyc-tlc.s3.amazonaws.com (nyc-tlc.s3.amazonaws.com)... 54.231.192.161
Connecting to nyc-tlc.s3.amazonaws.com (nyc-tlc.s3.amazonaws.com)|54.231.192.161|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 11886281 (11M) [binary/octet-stream]
Saving to: ‘data/fhv_tripdata_2021-01.parquet’

fhv_tripdata_2021-0 100%[===================>]  11.33M  2.75MB/s    in 5.6s    

2022-06-04 23:19:20 (2.04 MB/s) - ‘data/fhv_tripdata_2021-01.parquet’ saved [11886281/11886281]

--2022-06-04 23:19:20--  https://nyc-tlc.s3.amazonaws.com/trip+data/fhv_tripdata_2021-02.parquet
Reusing existing connection to nyc-tlc.s3.amazonaws.com:443.
HTTP request sent, awaiting response... 200 OK
Length: 10645466 (10M) [binary/octet-stream]
Saving to: ‘data/fhv_tripdata_2021-02.parquet’

fhv_tripdata_2021-0 100%[===================>]  10.15M  5.85MB/s    in 1.7s    

2022-06-04 23:19:22 (5.85 


## Q1. Converting the script to a Prefect flow

If you want to follow the videos exactly, do:

```
pip install prefect==2.0b5
```

If you need Windows support, check `homework-windows.md` for installation instructions.

The current script `homework.py` is a fully functional script as long as you already have `fhv_trip_data_2021-01.parquet` and `fhv_trip_data_2021-02.parquet` inside a `data` folder. You should be able to already run it using:

```
python homework.py
```

We want to bring this to workflow orchestration to add observability around it. The `main` function will be converted to a `flow` and the other functions will be `tasks`. After adding all of the decorators, there is actually one task that you will need to call `.result()` for inside the `flow` to get it to work. Which task is this?

* read_data
* prepare_features
* train_model
* run_model

Important: change all `print` statements to use the Prefect logger. Using the `print` statement will not appear in the Prefect UI. You have to call `get_run_logger` at the start of the task to use it.


In [2]:
!python homework.py

20:51:28.293 | INFO    | prefect.engine - Created flow run 'chirpy-mackerel' for flow 'main'
20:51:28.293 | INFO    | Flow run 'chirpy-mackerel' - Using task runner 'SequentialTaskRunner'
20:51:28.300 | WARNING | Flow run 'chirpy-mackerel' - No default storage is configured on the server. Results from this flow run will be stored in a temporary directory in its runtime environment.
20:51:28.339 | INFO    | Flow run 'chirpy-mackerel' - Created task run 'read_data-4c7f9de4-0' for task 'read_data'
20:51:30.890 | INFO    | Task run 'read_data-4c7f9de4-0' - Finished in state Completed()
20:51:30.912 | INFO    | Flow run 'chirpy-mackerel' - Created task run 'prepare_features-4ee39d9f-0' for task 'prepare_features'
20:51:30.999 | INFO    | Task run 'prepare_features-4ee39d9f-0' - The mean duration of training is 16.2472533682457
20:51:34.467 | INFO    | Task run 'prepare_features-4ee39d9f-0' - Finished in state Completed()
20:51:34.513 | INFO    | Flow run 'chirpy-mackerel' - Created task run

After adding all of the decorators, there is actually one task that you will need to call `.result()` for inside the `flow` to get it to work. Which task is this?

**--> C) train_model**

since the call to this task is unpacking into lr, dv = ...

## Q2. Parameterizing the flow

Right now there are two parameters for `main()` called `train_path` and `val_path`. We want to change the flow function to accept `date` instead. `date` should then be passed to a task that gives both the `train_path` and `val_path` to use.

It should look like this:

```python
@flow
def main(date=None):
    train_path, val_path = get_paths(date).result()
    # rest of flow below
```

Where `get_paths` is a task that you have to implement. The specs for this are outlined in the motivation section. Listing them out again here:

The flow will take in a parameter called `date` which will be a datetime.
    a. `date` should default to None
    b. If `date` is None, use the current day. Use the data from 2 months back as the training data and the data from the previous month as validation data.
    c. If a `date` value is supplied, get 2 months before the `date` as the training data, and the previous month as validation data.
    d. As a concrete example, if the date passed is "2021-03-15", the training data should be "fhv_tripdata_2021-01.parquet" and the validation file will be "fhv_trip_data_2021-02.parquet"

Because we have two files:
* fhv_tripdata_2021-01.parquet
* fhv_tripdata_2021-02.parquet

Change the `main()` flow call to the following:

```
main(date="2021-03-15")
```

and it should use those files. This is a simplification for testing our homework.

Recall the page from where we downloaded the For-Hire trip data.

https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page

Download the relevant files needed to run the `main` flow if `date` is 2021-08-15.

For example:
```
main(date="2021-08-15")
```

By setting up the logger from the previous step, we should see some logs about our training job. What is the validation MSE when running the flow with this date?

Note you need to download the relevant files to run. Part of this question is understanding which files the flow should be looking for.

The valition MSE is:

* 11.637
* 11.837
* 12.037
* 12.237

## Testing to find a solution for get_paths implementation
( local, in the notebook )

In [2]:
from datetime import date, timedelta, datetime
import dateutil.relativedelta


date_string = "2021-08-15"
d = date.fromisoformat(date_string)

training_months_dt = [d - dateutil.relativedelta.relativedelta(months=i) for i in range(1, 2)]
validation_months_dt = [d - dateutil.relativedelta.relativedelta(months=i) for i in range(2, 3)]

training_months_paths = [f'./data/fhv_tripdata_{dt.year}-{dt.month}.parquet' for dt in training_months_dt]
validation_months_dt = [f'./data/fhv_tripdata_{dt.year}-{dt.month}.parquet' for dt in validation_months_dt]

print(training_months_paths)
print(validation_months_dt)


['./data/fhv_tripdata_2021-7.parquet']
['./data/fhv_tripdata_2021-6.parquet']


In [3]:
from datetime import date, timedelta, datetime
import dateutil.relativedelta

def get_paths(date_string=None, months_back_for_training_range=(1, 2), months_back_for_valid_range=(2, 3)):
    if date_string is None:
        date_string = date.today()

    datem = date.fromisoformat(date_string)

    training_months_dt = [datem - dateutil.relativedelta.relativedelta(months=i) for i in months_back_for_training_range]
    validation_months_dt = [datem - dateutil.relativedelta.relativedelta(months=i) for i in months_back_for_valid_range]

    training_months_paths = [f'./data/fhv_tripdata_{dt.year}-{dt.month:02}.parquet' for dt in training_months_dt]
    validation_months_paths = [f'./data/fhv_tripdata_{dt.year}-{dt.month:02}.parquet' for dt in validation_months_dt]

    return training_months_paths[0], validation_months_paths[0] # return single files for training, since main()  expects it in this format


assert(get_paths("2021-03-15") == ('./data/fhv_tripdata_2021-02.parquet', './data/fhv_tripdata_2021-01.parquet'))

get_paths("2021-03-15")

('./data/fhv_tripdata_2021-02.parquet', './data/fhv_tripdata_2021-01.parquet')

In [29]:
#Download the relevant files needed to run the `main` flow if `date` is 2021-08-15.

!wget -P data 'https://nyc-tlc.s3.amazonaws.com/trip+data/fhv_tripdata_2021-0'{6..7}'.parquet'

--2022-06-06 13:17:29--  https://nyc-tlc.s3.amazonaws.com/trip+data/fhv_tripdata_2021-06.parquet
Resolving nyc-tlc.s3.amazonaws.com (nyc-tlc.s3.amazonaws.com)... 52.216.160.3
Connecting to nyc-tlc.s3.amazonaws.com (nyc-tlc.s3.amazonaws.com)|52.216.160.3|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 13208079 (13M) [binary/octet-stream]
Saving to: ‘data/fhv_tripdata_2021-06.parquet’

fhv_tripdata_2021-0 100%[===================>]  12.60M  2.82MB/s    in 13s     

2022-06-06 13:17:43 (1018 KB/s) - ‘data/fhv_tripdata_2021-06.parquet’ saved [13208079/13208079]

--2022-06-06 13:17:43--  https://nyc-tlc.s3.amazonaws.com/trip+data/fhv_tripdata_2021-07.parquet
Reusing existing connection to nyc-tlc.s3.amazonaws.com:443.
HTTP request sent, awaiting response... 200 OK
Length: 12650862 (12M) [binary/octet-stream]
Saving to: ‘data/fhv_tripdata_2021-07.parquet’

fhv_tripdata_2021-0 100%[===================>]  12.06M  5.40MB/s    in 2.2s    

2022-06-06 13:17:46 (5.40 MB/s

In [38]:
# Run the homework for hw3.Q2
# main(date="2021-08-15")
!python homework.py

13:38:08.129 | INFO    | prefect.engine - Created flow run 'neat-puffin' for flow 'main'
13:38:08.129 | INFO    | Flow run 'neat-puffin' - Using task runner 'SequentialTaskRunner'
13:38:08.136 | WARNING | Flow run 'neat-puffin' - No default storage is configured on the server. Results from this flow run will be stored in a temporary directory in its runtime environment.
13:38:08.171 | INFO    | Flow run 'neat-puffin' - Created task run 'get_paths-6e696e34-0' for task 'get_paths'
13:38:08.209 | INFO    | Task run 'get_paths-6e696e34-0' - Finished in state Completed()
13:38:08.210 | INFO    | Flow run 'neat-puffin' - Path for training: ./data/fhv_tripdata_2021-06.parquet, Path for validation: ./data/fhv_tripdata_2021-07.parquet
13:38:08.230 | INFO    | Flow run 'neat-puffin' - Created task run 'read_data-4c7f9de4-0' for task 'read_data'
13:38:11.192 | INFO    | Task run 'read_data-4c7f9de4-0' - Finished in state Completed()
13:38:11.225 | INFO    | Flow run 'neat-puffin' - Created task r

The valition MSE is:

* 11.637
* 11.837
* 12.037
* 12.237

13:38:32.315 | INFO    | Task run 'run_model-6559300c-0' - The MSE of validation is: 11.637023826050765

**--> A) 11.637**


## Q3. Saving the model and artifacts

At the moment, we are not saving the model and vectorizer for future use. You don't need a new task for this, you can just add it inside the `flow`. The requirements for filenames to save it as were mentioned in the Motivation section. They are pasted again here:

* Save the model as "model-{date}.pkl" where date is in `YYYY-MM-DD`. Note that `date` here is the value of the flow `parameter`. In practice, this setup makes it very easy to get the latest model to run predictions because you just need to get the most recent one.
* In this example we use a DictVectorizer. That is needed to run future data through our model. Save that as "dv-{date}.pkl". Similar to above, if the date is `2021-03-15`, the files output should be `model-2021-03-15.bin` and `dv-2021-03-15.b`.

By using this file name, during inference, we can just pull the latest model from our model directory and apply it. Assuming we already had a list of filenames:

```python
['model-2021-03-15.bin', 'model-2021-04-15.bin', 'model-2021-05-15.bin']
```

We could do something like `sorted(model_list, reverse=False)[0]` to get the filename of the latest file. This is the simplest way to consistently use the latest trained model for inference. Tools like MLFlow give us more control logic to use flows.

What is the file size of the `DictVectorizer` that we trained when the `date` is 2021-08-15?

* 13,000 bytes
* 23,000 bytes
* 33,000 bytes
* 43,000 bytes

You can bring up the Orion UI and see the work you've been doing. If you are using local Orion, you can start the server with:

```bash
prefect orion start
```

You should be able to see previous Flow runs and the most recent successful runs. Navigate to some of them. Take time to explore the UI. The radar plot would be a good thing to share on social media if you participate in those posts.



In [39]:
!mkdir models

Add saving the model and dictVectorizer into def main(date): function,
after training is done:

```
lr, dv = train_model(df_train_processed, categorical).result()

with open(f'models/model-{date}.pkl', "wb") as f_out:
    pickle.dump(lr, f_out)
with open(f'models/dv-{date}.pkl', "wb") as f_out:
    pickle.dump(dv, f_out)
```

In [49]:
# Run the homework for hw3.Q3
# main(date="2021-08-15")
!python homework.py

15:14:46.970 | INFO    | prefect.engine - Created flow run 'wondrous-manatee' for flow 'main'
15:14:46.970 | INFO    | Flow run 'wondrous-manatee' - Using task runner 'SequentialTaskRunner'
15:14:46.977 | WARNING | Flow run 'wondrous-manatee' - No default storage is configured on the server. Results from this flow run will be stored in a temporary directory in its runtime environment.
15:14:47.011 | INFO    | Flow run 'wondrous-manatee' - Created task run 'get_paths-6e696e34-0' for task 'get_paths'
15:14:47.046 | INFO    | Task run 'get_paths-6e696e34-0' - Finished in state Completed()
15:14:47.046 | INFO    | Flow run 'wondrous-manatee' - Path for training: ./data/fhv_tripdata_2021-06.parquet, Path for validation: ./data/fhv_tripdata_2021-07.parquet
15:14:47.060 | INFO    | Flow run 'wondrous-manatee' - Created task run 'read_data-4c7f9de4-0' for task 'read_data'
15:14:49.932 | INFO    | Task run 'read_data-4c7f9de4-0' - Finished in state Completed()
15:14:49.951 | INFO    | Flow run 

In [53]:
!du -b -c ./models/*
#!du -bch ./models

13191	./models/dv-2021-08-15.pkl
4581	./models/model-2021-08-15.pkl
17772	total


What is the file size of the `DictVectorizer` that we trained when the `date` is 2021-08-15?

* 13,000 bytes
* 23,000 bytes
* 33,000 bytes
* 43,000 bytes

**--> A) 13,000 bytes**
>13191	./models/dv-2021-08-15.pkl


## Q4. Creating a deployment with a CronSchedule

We previously showed the `IntervalSchedule` in the video tutorials. In some cases, the interval is too rigid. For example, what if we wanted to run this `flow` on the 15th of every month? An interval of 30 days would not be in sync. In cases like these, the `CronSchedule` is more appropriate. The documentation for that is [here](https://orion-docs.prefect.io/concepts/schedules/#cronschedule)

Cron is an important part of workflow orchestration. It is used to schedule tasks, and was a predecessor for more mature orchestration frameworks. A lot of teams still use Cron in production. Even if you don't use Cron, the Cron expression is very common as a way to write a schedule, and the basics are worth learning for orchestration, even outside Prefect.

For this exercise, use a `CronSchedule` when creating a Prefect deployment.

What is the Cron expression to run a flow at 9 AM every 15th of the month?

* `* * 15 9 0`
* `9 15 * * *`
* `0 9 15 * *`
* `0 15 9 1 *`

Hint: there are many Cron to English tools. Try looking for one to help you.

Create a deployment with `prefect deployment create` after you write your `DeploymentSpec`


In [ ]:
!prefect deployment create homework.py

[https://crontab.guru/#0_9_15_*_*](https://crontab.guru/#0_9_15_*_*)
Answer: `0 9 15 * *`

**--> C)**

## Q5. Viewing the Deployment

View the deployment in the UI. When first loading, we may not see that many flows because the default filter is 1 day back and 1 day forward. Remove the filter for 1 day forward to see the scheduled runs.

How many flow runs are scheduled by Prefect in advanced? You should not be counting manually. There is a number of upcoming runs on the top right of the dashboard.

* 0
* 3
* 10
* 25

Answer: 3
There are 3 scheduled flow runs in advance.

**--> B) 3**


## Q6. Creating a work-queue

In order to run this flow, you will need an agent and a work queue. Because we scheduled our flow or every month, it won't really get picked up by an agent. For this exercise, create a work-queue from the UI and view it using the CLI.

For all CLI commands with Prefect, you can use `--help` to get more information.

For example,
* `prefect --help`
* `prefect work-queue --help`

What is the command to view the available work-queues?

* `prefect work-queue inspect`
* `prefect work-queue ls`
* `prefect work-queue preview`
* `prefect work-queue list`


In [57]:
!prefect work-queue ls

                             Work Queues                             
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃                                   ID ┃ Name   ┃ Concurrency Limit ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ 3c3b2278-2065-4f40-85f9-7b0e704817d4 │ hw3-q6 │ None              │
└──────────────────────────────────────┴────────┴───────────────────┘
                     (**) denotes a paused queue                     


Answer:

**--> B) prefect work-queue ls**
